clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -2669,6 +2742,96 @@ scheduler_tmpl_t<RecordType, ReaderType>::process_marker(input_info_t &input,
         } else {
             input.switch_to_input = it->second;
         }
+        // Trigger a switch either indefinitely or until timeout.
+        if (input.skip_next_unscheduled) {
+            input.skip_next_unscheduled = false;
+            VPRINT(this, 3,
+                   "input %d unschedule request ignored due to prior schedule request "
+                   "@%" PRIu64 "\n",
+                   input.index, input.reader->get_last_timestamp());
+            break;
+        }
+        input.unscheduled = true;
+        if (input.syscall_timeout_arg > 0) {
+            input.blocked_time = scale_blocked_time(input.syscall_timeout_arg);
+            VPRINT(this, 3, "input %d unscheduled for %" PRIu64 " @%" PRIu64 "\n",
+                   input.index, input.blocked_time, input.reader->get_last_timestamp());
+        } else {
+            VPRINT(this, 3, "input %d unscheduled indefinitely @%" PRIu64 "\n",
+                   input.index, input.reader->get_last_timestamp());
+        }
+        break;
+    }
+    case TRACE_MARKER_TYPE_SYSCALL_ARG_TIMEOUT:
+        // This is cleared at the post-syscall instr.
+        input.syscall_timeout_arg = static_cast<uint64_t>(marker_value);
+        break;
+    case TRACE_MARKER_TYPE_SYSCALL_UNSCHEDULE:
+        if (!options_.honor_direct_switches)
+            break;
+        if (input.skip_next_unscheduled) {
```

Comment:
Explain this further. It's interesting that a prior "schedule" marker may have such an effect on a future "unschedule", especially because the "schedule" marker docs say "if it were __currently__ unschedulable".

What real world scenario is this trying to model? What if the "do-not-unschedule" input was blocked on a I/O syscall?

Also, (not saying that we should) what's the rationale behind skip_next_unscheduled being a bool and not an int (maybe multiple other threads want to "schedule" the thread), therefore the thread should get that many get-out-of-unschedulable-state tokens.

Reply:
The real-world kernel allows only a single skip. I believe the real-world feature's rationale is to avoid the need for synchronization in some not-uncommon pattern of calls but I'm not sure I know what that pattern is.

Reply:
> The real-world kernel allows only a single skip.

Multiple threads could want some blocked thread to be scheduled, perhaps at different times in the app's execution, but at the same time when the trace is simulated/scheduled.

Reply:
> especially because the "schedule" marker docs say "if it were currently unschedulable".

This was updated, as noted in the other thread.

> Multiple threads could want some blocked thread to be scheduled, perhaps at different times in the app's execution, but at the same time when the trace is simulated/scheduled.

Maybe.  There are many theoretical cases.  This re-scheduling is always going to be best-effort.  Trying to have a queue of wakers seems just as likely to cause non-representative behavior than to fix it though so it does not seem worth going in that direction.  Added a comment.

Reply:
If the "schedule" events get lost due to getting accumulated, then the target thready may experience a hang if some of its "unschedulable" events don't get matched up with a "schedule".